docs: revise CLI usage documentation (#147)
mergecal/icalendar-anonymizer: docs/usage/cli.rst

Reviewer: ````suggestion
     - When used
````

Sentence case per MS Style Guide.

Reviewer: For 88, 262, 268, 283, 291, 302, 308, 317:

Most shell examples assume a Unix-like environment (`cat`, `grep`, `for file in *.ics; do`, `while read`, `curl | ...`). Windows users on cmd or PowerShell hit different syntax. The "Binary mode on Windows" section on line 469 covers one PowerShell pattern but not the broader pipeline/batch examples.

I'd lean toward keeping Unix as the primary path (matches where we're heading) but adding a short note at the top of "Pipeline processing" and "Batch processing" stating the examples assume Bash, with a pointer to the Windows section. What say?

Author: First, I agree that Unix should be primary path.

Second, I think it's up to you what you and Click want to support. From what I could pull from the Click docs, it appears that they favor `cmd` over PowerShell. Source: https://click.palletsprojects.com/en/stable/search/?q=windows.

I don't have Windows, so I can't test the flavors of its shell commands. We would also need to adapt commands with paths in their names to use `\` instead of `/` and check path names, such as `\Scripts`.

Reviewer: Agreed Unix as primary. PowerShell for the Windows section per https://github.com/pycalendar/icalendar-anonymizer/pull/147#discussion_r3372885310. 

I'll handle the Windows-side testing.

Author: I added an example of Sphinx Design tabs for activating the Python virtual environment, which you can replicate for all the PowerShell commands. That way you don't need to split the usage docs for the users's OS.

.. SPDX-FileCopyrightText: 2025 icalendar-anonymizer contributors
.. SPDX-License-Identifier: AGPL-3.0-or-later

**********************
Command-line interface
**********************

The command-line interface (CLI) provides Unix-style tools for anonymizing iCalendar files from the terminal.

Installation
============

First, install uv as described in :ref:`development-prerequisites`.

Install the latest Python version and create virtual environment.

..  code-block:: shell

    uv venv

Activate the virtual environment.

..  code-block:: shell

>     source .venv/bin/activate

Reviewer: The install flow uses `source .venv/bin/activate`, which is Unix-only. Windows users need `.venv\Scripts\Activate.ps1` (PowerShell) or `.venv\Scripts\activate.bat` (cmd).

Consistent with the "Unix primary path" decision, but for the very first install step it's worth a one-liner pointing Windows users at the right command. Could be a `.. note::` block right after line 25:

```rst
..  note::

    On Windows, activate the venv with ``.venv\Scripts\Activate.ps1`` (PowerShell)
    or ``.venv\Scripts\activate.bat`` (cmd).
```

Author: Updated with Sphinx Design tabs in 4dcdea12477deaa13f359c7a5f5f63e3a7472548

    uv pip install "icalendar-anonymizer[cli]"

Verify installation with the following command.

..  code-block:: shell

    ican --version

This should output the package name and its version number.

Commands
========

Two commands are provided.

:program:`icalendar-anonymize`
    Full command name
:program:`ican`
    Short alias for convenience

Both commands work identically.

Basic usage
===========

This section describes how to use the command-line :program:`icalendar-anonymize` application.
For brevity, examples use the alias form.

The basic usage syntax calls the program, followed optionally by options, then input, and finally output.

..  code-block:: shell

    ican [OPTIONS] [INPUT] [-o OUTPUT]

Anonymize a file
----------------

Read from a file and write to another file:

..  code-block:: shell

    ican calendar.ics -o anonymized.ics

Write to ``stdout``
-------------------

Omit the ``-o`` flag to write to ``stdout``:

..  code-block:: shell

    ican calendar.ics

Read from ``stdin``
-------------------

Omit the input argument to read from ``stdin``:

..  code-block:: shell

    cat calendar.ics | ican > anonymized.ics

Field configuration examples

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ````suggestion
Field configuration examples
----------------------------
````

After the Options reference move, `Field configuration examples` is sitting at heading level 4 (`^^^^`) under "Read from stdin" instead of level 3 (`----`) as a sibling of "Anonymize a file" / "Write to stdout" / "Read from stdin". Renders as a subsection of stdin handling, which is wrong?

Author: Actually, I decided to merge the two usage sections into just "Usage", and remove duplicate examples. Now it flows better. See d54c295eaef59d026cbd669031d65350a7d2bc4c.

        
            source .venv/bin/activate
        
    ..  tab-item:: Windows PowerShell

        ..  code-block:: powershell
        
            .venv\Scripts\Activate.ps1
        
Install icalendar-anonymizer, its dependencies, and the CLI extras.

..  code-block:: shell

    uv pip install "icalendar-anonymizer[cli]"

Verify installation with the following command.

..  code-block:: shell

    ican --version

This should output the package name and its version number.

Commands
========

Two commands are provided.

:program:`icalendar-anonymize`
    Full command name
:program:`ican`
    Short alias for convenience

Both commands work identically.

Usage
=====

This section describes how to use the command-line :program:`icalendar-anonymize` application.
For brevity, examples use the alias form.

The usage syntax calls the program, followed optionally by options, then input, and finally output.

..  code-block:: shell

    ican [OPTIONS] [INPUT] [-o OUTPUT]

For a complete list of command options and field configuration options, either see :ref:`options-reference` or :ref:`field-configuration-options`, or use the following command.

..  code-block:: shell

>     make --help

Reviewer: ````suggestion
    ican --help
````

Says `make --help`. Should be `ican --help`. Copy-paste artifact from the merge of the two usage sections in d54c295. The same `ican --help` example shows up correctly on line 122 in the Field configuration section.

Reviewer: Doen in [f182bda](https://github.com/pycalendar/icalendar-anonymizer/pull/147/commits/f182bda53a427f6bea0189407ed8d2206935630e)

Author: What was I thinking? I got `make` on the brain. Careful, it might be contagious.

Thanks for fixing this.

.. SPDX-FileCopyrightText: 2025 icalendar-anonymizer contributors
.. SPDX-License-Identifier: AGPL-3.0-or-later

**********************
Command-line interface
**********************

The command-line interface (CLI) provides Unix-style tools for anonymizing iCalendar files from the terminal.

Installation
============

First, install uv as described in :ref:`development-prerequisites`.

Then, clone icalendar-anonymizer and change your working directory to the root of the repository.

..  code-block:: shell

    git clone https://github.com/pycalendar/icalendar-anonymizer.git
    cd icalendar-anonymizer

Then, install the CLI from a local checkout of icalendar-anonymizer with the following command.

..  code-block:: shell

    uv pip install --group cli

> This installs the Click dependency and both command-line tools.

Reviewer: ````suggestion
Then, install icalendar-anonymizer with its CLI dependencies from a local checkout with the following command.

..  code-block:: shell

    uv pip install -e . --group cli

This installs the package itself plus the Click dependency, exposing both command-line tools.
````

`uv pip install --group cli` installs the dependency group's contents (just Click) but not the package itself:

```
$ uv pip install --group cli --dry-run
Would install 2 packages
 + click==8.4.1
 + colorama==0.4.6
```

`ican --version` on the next step fails with `command not found`. Adding `-e .` installs the package alongside the group.

Separate question: is clone-and-install really the right path for CLI users? Cloning is a contributor workflow. `pip install icalendar-anonymizer[cli]` from PyPI is more typical for end users once we cut a release. Up to you whether to keep this, swap to PyPI, or document both?

Author: Yeah, cloning is not the right path for this audience. I'll adapt the commands, now that we can use a mixture of `--group` and `--extra` for uv.

Author: Actually, see the change in d85b6113087b1787b33afa8464f2eff31945401e. It's a lot better.

Reviewer: Sounds good. The new `uv pip install "icalendar-anonymizer[cli]"` works end-to-end, I verified against PyPI 0.1.4. One downstream nit on the troubleshooting tip below (still uses `uv sync`).

Verify installation with the following command.

..  code-block:: shell

    ican --version

This should output the package name and its version number.

Commands
========

Two commands are provided.

:program:`icalendar-anonymize`
    Full command name
:program:`ican`
    Short alias for convenience

Both commands work identically.

Basic usage
===========

This section describes how to use the command-line :program:`icalendar-anonymize` application.
For brevity, examples use the alias form.

The basic usage syntax calls the program, followed optionally by options, then input, and finally output.

..  code-block:: shell

    ican [OPTIONS] [INPUT] [-o OUTPUT]

Anonymize a file
----------------

Read from a file and write to another file:

..  code-block:: shell

    ican calendar.ics -o anonymized.ics

Write to ``stdout``
-------------------

Omit the ``-o`` flag to write to ``stdout``:

..  code-block:: shell

    ican calendar.ics

Read from ``stdin``
-------------------

Omit the input argument to read from ``stdin``:

..  code-block:: shell

    cat calendar.ics | ican > anonymized.ics

Alternatively, use ``-``. 

..  code-block:: shell

>     ican - -o anonymized.ics

Reviewer: ````suggestion
````

Drop it. `ican - -o anonymized.ics` without piped stdin blocks waiting for input in an interactive terminal. Line 88 already covers stdin clearly.

If you want explicit `-` documented, line 268 already has it in the Pipeline processing section using input redirection.

Author: GitHub didn't like this suggestion either. Will do locally.

Author: Removed in 1ad6aa94604a9d80e4262dac408ff6425c07c867.

Options reference
=================

..  program:: icalendar-anonymize
 
..  option:: [INPUT]
 
    Input iCalendar file to anonymize.
    Optional positional argument.
 
    -   **Default**: ``stdin`` (``-``)
    -   **Format**: File path or ``-`` for ``stdin``
    -   **Example**: :code:`ican calendar.ics`
 
..  option:: -o <file>, --output <file>
 
    Output file for anonymized calendar.
 
    -   **Default**: ``stdout`` (``-``)
    -   **Format**: File path or ``-`` for ``stdout``
    -   **Example**: :code:`ican input.ics -o output.ics`
 
..  option:: -v, --verbose
 
    Show processing information on stderr. Displays input/output sources and processing steps.
 
    -   **Flag**: No value required
    -   **Output**: Messages written to stderr (not ``stdout``)
    -   **Example**: :code:`ican -v calendar.ics -o anonymized.ics`
 
    The following example shows verbose output:
 
    ..  code-block:: text
 
        Reading from: calendar.ics
        Parsing calendar...
        Anonymizing calendar...
        Writing to: anonymized.ics
        Done.

..  option:: --version
 
    Display version information and exit.
 
    ..  code-block:: shell
 
        $ ican --version

>         icalendar-anonymizer, version 0.1.0

Reviewer: ````suggestion
        $ ican --version
        icalendar-anonymizer, version <version>
````

`0.1.0` is hardcoded but `hatch-vcs` derives version from git tags. Goes stale the moment we cut a release.

 
..  option:: --help
 
    Show usage information and exit.
 
    ..  code-block:: shell
 
        ican --help

    ..  note::

>         The output for "Usage" is somewhat misleading, as Click merges ``-o, --output FILENAME`` with the options instead of as a positional final optional argument.

Reviewer: Worth cross-referencing #148 since that's also about confusing Click `--help` output (the epilog example wrap):

```
.. note::

    The output for "Usage" is somewhat misleading, as Click merges ``-o, --output FILENAME``
    with the options instead of as a positional final optional argument. See also :issue:`148`
    for a related Click formatting quirk.
```

Optional :)

Author: ```suggestion
        The output for "Usage" is somewhat misleading, as Click merges ``-o, --output FILENAME`` with the options instead of as a positional final optional argument.
        See also :issue:`148` for a related Click formatting quirk.
```

..  code-block:: shell

    cat calendar.ics | ican > anonymized.ics

Explicitly read from ``stdin`` with ``-`` and write to ``stdout``.

..  code-block:: shell

    ican - < calendar.ics > anonymized.ics

Verbose output to ``stderr`` doesn't corrupt ``stdout``.

..  code-block:: shell

    cat calendar.ics | ican -v > anonymized.ics

Batch processing
----------------

Anonymize all ICS files in directory.

..  code-block:: shell

    for file in *.ics; do
        ican "$file" -o "anonymized-$file"
    done

Process files from a list.

..  code-block:: shell

    while read -r file; do
        ican "$file" -o "anon-$(basename "$file")"
    done < file-list.txt

Remote files
------------

Download a remote file and anonymize it.

..  code-block:: shell

    curl https://example.com/calendar.ics | ican > local-anon.ics

Do the previous example with error checking.

..  code-block:: shell

    curl -f https://example.com/calendar.ics | ican -v > local-anon.ics

Combining with other tools
--------------------------

Anonymize and count events.

..  code-block:: shell

    ican calendar.ics | grep -c "BEGIN:VEVENT"

Anonymize and validate the input.

..  code-block:: shell

>     ican calendar.ics | ics-validator

Reviewer: ````suggestion
    ican calendar.ics | <your-validator>
````

`ics-validator` isn't a real package on PyPI. Reads like one, so a placeholder is clearer.

    ican --summary keep --location remove --description replace calendar.ics

Pipeline processing
-------------------

..  note::

    PowerShell examples use the ``-o`` flag instead of ``>``. PowerShell 5.1 encodes ``>`` as UTF-16, which corrupts the output for tools that expect UTF-8.

    PowerShell 5.1 also re-encodes ``stdin`` pipes through the console code page and can mangle non-ASCII content. For data with accents or non-Latin text on 5.1, prefer the file-argument form (``ican calendar.ics -o anonymized.ics``). PowerShell 7+ handles UTF-8 through pipes correctly.

Read from a file and write to a file.

..  tab-set::

    ..  tab-item:: Linux and macOS
        :sync: linux

        ..  code-block:: shell

            cat calendar.ics | ican > anonymized.ics

    ..  tab-item:: Windows PowerShell
        :sync: powershell

        ..  code-block:: powershell

            ican calendar.ics -o anonymized.ics

Read from ``stdin`` explicitly with ``-``.

..  tab-set::

    ..  tab-item:: Linux and macOS
        :sync: linux

        ..  code-block:: shell

            ican - < calendar.ics > anonymized.ics

    ..  tab-item:: Windows PowerShell
        :sync: powershell

>         PowerShell pipes re-encoded bytes through the console code page.

Author: I couldn't tell whether "pipes" was a verb or a plural noun, but it felt more like a verb to me, so I changed "re-encode" from a verb to an adjective as "re-encoded" to align with my interpretation. If it was in fact a plural noun, then I'd preface the sentence with `The ` and revert this change.

Reviewer: Plural noun was what I meant. Even cleaner option I think is to rephrase to remove the ambiguity entirely.

```suggestion
        PowerShell re-encodes piped bytes through the console code page. On 5.1 this corrupts non-ASCII characters, even with `-Encoding utf8`. Pass the file path directly instead.
```

What do you think?

Author: I reused phrasing from the note above. See 3893c9e640e89e2feac42d5fda1fa42467a3f0de

    cd icalendar-anonymizer

Then, install the CLI from a local checkout of icalendar-anonymizer with the following command.

..  code-block:: shell

    uv pip install --group cli

This installs the Click dependency and both command-line tools.

Verify installation with the following command.

..  code-block:: shell

    ican --version

This should output the package name and its version number.

Commands
========

Two commands are provided.

:program:`icalendar-anonymize`
    Full command name
:program:`ican`
    Short alias for convenience

Both commands work identically.

Basic usage
===========

This section describes how to use the command-line :program:`icalendar-anonymize` application.
For brevity, examples use the alias form.

Anonymize a file
----------------

Read from a file and write to another file:

.. code-block:: shell

    ican calendar.ics -o anonymized.ics

Write to ``stdout``
-------------------

Omit the ``-o`` flag to write to ``stdout``:

.. code-block:: shell

    ican calendar.ics

Read from ``stdin``
-------------------

Omit the input argument, or use ``-``, to read from ``stdin``:

.. code-block:: shell

    cat calendar.ics | ican > anonymized.ics

>     ican - -o anonymized.ics

Author: This command hung. I'm not sure we should include it.

Reviewer: Confirmed removed in 1ad6aa9. Resolving.

    # Anonymize and compress
    cat calendar.ics | ican | gzip > anonymized.ics.gz

> Options reference

Author: By virtue of its name, this is section and its subsection are reference material, not a how-to guide. I'd move them accordingly.

Reviewer: Yes. I'd lean toward option 1: pull it into a new `docs/reference/cli.rst` and link from here. Option 2 (move to the bottom of this page) is the smallest change if you want to keep the scope tight for this PR.

Author: Moved temporarily to the end with a todo, and created https://github.com/pycalendar/icalendar-anonymizer/issues/153.

     - Input file not found or cannot be opened
   * - 130
     - Interrupted
     - User pressed :kbd:`Ctrl+C` (SIGINT)

Troubleshooting
===============

The following sections provide troubleshooting tips.

Command not found
-----------------

If you get ``command not found`` after installation:

#.  Reinstall with the CLI dependency group:
 
    ..  code-block:: shell
 

>         uv sync --group cli

Reviewer: ````suggestion
#.  Reinstall the package with the CLI extras:
 
    ..  code-block:: shell
 
        uv pip install --reinstall "icalendar-anonymizer[cli]"
````

The troubleshooting tip uses `uv sync --group cli` which is a project-level uv command. The install path on line 31 is `uv pip install "icalendar-anonymizer[cli]"`, so the user isn't in a uv-managed project and `uv sync` won't work for them. The reinstall should match the install.

Author: Good catch. Fixed!

**Exit code**: 2

> Invalid ICS file

Author: This heading and its error message should be `ics`. ICS is an acronym, whereas `ics` is the file extension of iCalendar files. Better yet, spell it out to "iCalendar".

Leaving as is for now.

Reviewer: I'd actually go further and rename to `Invalid iCalendar file` everywhere, including [the error message in `cli.py`](https://github.com/pycalendar/icalendar-anonymizer/blob/17177f8bd580e757084964583d47e51aeabb7b7b/src/icalendar_anonymizer/cli.py#L147). Matches how the rest of the docs spell out iCalendar and avoids the acronym-vs-extension ambiguity entirely. Fine to defer to a follow-up issue if you want to keep this PR focused.

Author: I had the same thought for error messages, but didn't want to push it in here. Let's defer this one. Issue created https://github.com/pycalendar/icalendar-anonymizer/issues/152.

Reviewer: Thanks for filing the issue!

     - Input file not found or cannot be opened
   * - 130
     - Interrupted
     - User pressed :kbd:`Ctrl+C` (SIGINT)

Troubleshooting
===============

The following sections provide troubleshooting tips.

Command not found
-----------------

If you get ``command not found`` after installation:

#.  Reinstall with the CLI dependency group:
 
    ..  code-block:: shell
 
        uv sync --group cli
 
#.  Use the full Python module path:
 
    ..  code-block:: shell
 
        python -m icalendar_anonymizer.cli calendar.ics

Binary mode on Windows
----------------------

The CLI automatically handles binary mode on Windows.
You don't need to worry about CRLF line endings.

If you encounter encoding issues on Windows, then use binary mode with PowerShell. 

> ..  code-block:: shell

Reviewer: ````suggestion
..  code-block:: powershell

    Get-Content calendar.ics -Raw | ican > anonymized.ics
````

Lexer should be `powershell`, matching the `ps1con` tab on line 534.

Side note: "Binary mode on Windows" doesn't fit what `Get-Content -Raw` does. `-Raw` reads as one string, not binary. Up to you whether to retitle.

Author: Let's revisit, after a decision of `cmd` vs. PowerShell.

Reviewer: I'd lean PowerShell for the Windows path. A few reasons:

1. Windows PowerShell 5.1 ships preinstalled on both Windows 10 and 11. cmd is also there, but Windows Terminal (the modern terminal app shipped with Windows 11) defaults to PowerShell as its profile.
2. Microsoft's own docs (Learn, the PowerShell scripting guide) lead with PowerShell for new content, so it's where Windows users coming via docs end up.
3. PowerShell Core 7 runs cross-platform (Linux, macOS too), so the cmdlets we'd document are at least familiar shape to Unix users skimming the page.

cmd's still a fine fallback for anyone who prefers it but I wouldn't make it the documented path. I'm on Windows so I can test the snippets before merge. 

Plan: fix the lexer to `powershell` on line 356, and add a short PowerShell variant next to Unix pipeline/batch examples where syntax really differs. I can push a follow-up commit once you've landed the rest.

Author: Sounds reasonable to me. I'll use tabs from sphinx-design.

    Get-Content calendar.ics -Raw | ican > anonymized.ics

Large files
-----------

The CLI loads the entire file into memory.
For large files over 100MB in size, the following tips will improve performance.

-   Monitor memory usage.
    Use verbose mode to track progress.

    ..  code-block:: shell

        ican -v large-file.ics -o output.ics

-   Process in chunks.
    Split large calendars before anonymizing.

    The following example splits the calendar by year, then anonymizes it.

    ..  code-block:: shell

>         grep -A 100 "DTSTART:2024" calendar.ics | ican > 2024-anon.ics

Reviewer: ````suggestion
-   Use the :doc:`python-api` for programmatic control over memory usage and splitting by date range.
````

The `grep` example produces invalid ICS:

```
$ grep -A 100 "DTSTART:2025" tiny.ics | ican
Error: Invalid ICS file - Property "DTSTART" does not have a parent component.
```

`grep -A` doesn't preserve VCALENDAR/VEVENT wrappers. There isn't a clean shell one-liner for this (DTSTART can have TZID params, VALUE=DATE variants, line continuations, etc.), so I'd just collapse "Process in chunks" into the third bullet on line 503 which already points at the Python API. Anyone hitting a 100MB ICS file probably needs the Python route anyway.

If you want to keep a chunking example in the docs, the Python snippet below works and is short enough to drop inline:

```python
from icalendar import Calendar
from icalendar_anonymizer import anonymize

cal = Calendar.from_ical(open("calendar.ics").read())
year_cal = Calendar()
for event in cal.walk("VEVENT"):
    if event["DTSTART"].dt.year == 2024:
        year_cal.add_component(event)
print(anonymize(year_cal).to_ical().decode())
```

But probably better to keep it in python-api.rst as a worked example and link from here.

Author: Let's direct users to the best place in the Python API.

Author: I couldn't commit this one from GitHub. Will try from local copy.

Author: - Removed invalid example in 8d45a1528b44e9ccffc8838c9c3e247207c217f3
- Created new issue to capture the code example https://github.com/pycalendar/icalendar-anonymizer/issues/154

    See :issue:`153`.

Options reference
=================

..  program:: icalendar-anonymize
 
..  option:: [INPUT]
 
    Input iCalendar file to anonymize.
    Optional positional argument.
 
    -   **Default**: ``stdin`` (``-``)
    -   **Format**: File path or ``-`` for ``stdin``
    -   **Example**: :code:`ican calendar.ics`
 
..  option:: -o <file>, --output <file>
 
    Output file for anonymized calendar.
 
    -   **Default**: ``stdout`` (``-``)
    -   **Format**: File path or ``-`` for ``stdout``
    -   **Example**: :code:`ican input.ics -o output.ics`
 
..  option:: -v, --verbose
 
    Show processing information on stderr. Displays input/output sources and processing steps.
 
    -   **Flag**: No value required
    -   **Output**: Messages written to stderr (not ``stdout``)
    -   **Example**: :code:`ican -v calendar.ics -o anonymized.ics`
 
    The following example shows verbose output:
 
    ..  code-block:: text
 
        Reading from: calendar.ics
        Parsing calendar...
        Anonymizing calendar...
        Writing to: anonymized.ics
        Done.

..  option:: --version
 
    Display version information and exit.
 
    ..  code-block:: shell

>  
>         $ ican --version
>         $ ican --version
>         icalendar-anonymizer, version <version>

Reviewer: ````suggestion
    ..  code-block:: shell
 
        $ ican --version
        icalendar-anonymizer, version <version>
````

Looks like a bad merge of the earlier suggestion, it's duplicated

Author: Fixed in 7dba64f76af4d0ae1fa5218cc3adf8293ac48dbc
